### PR TITLE
Remove deprecated arguments from API and clarify model_name and chat_template_name

### PR DESCRIPTION
**Pull request by lvhan028**
- remove `get_hf_config_content`. Use `get_model_arch` instead
- remove deprecated argument such as `tp`, `meta_instruction`, `cap` and so on
- remove deprecated functions in CLI, such as `kv_qparams`
- tell "model_name" and "chat_template_name" apart. "model_name" means the name of the served model, which is used by the api_server "v1/models". "chat_template_name" means the name of the chat template
- "--chat-template" can be a builtin chat template name of a customized json file

### CODE REVIEW

**Comment by zhulinJulia24** [comment]
internlm2-chat-7b w8a8 is not supported in this pr
reproduce step is:
1. do w8a8 quantization
`lmdeploy lite smooth_quant /nvme/qa_test_models/internlm/internlm2-chat-7b --work-dir /nvme/qa_test_models/internlm/internlm2-chat-7b-w8a8 --batch-size 32`

2. run pipeline script
```
from lmdeploy import (GenerationConfig, PytorchEngineConfig,
                      TurbomindEngineConfig, pipeline)
from lmdeploy.pytorch.engine import Engine
from transformers import AutoTokenizer


engine_config = TurbomindEngineConfig(model_name=None, model_format=None, tp=1, session_len=None, max_batch_size=128, cache_max_entry_count=0.8, cache_block_seq_len=64, enable_prefix_caching=False, quant_policy=4, rope_scaling_factor=0.0, use_logn_attn=False, download_dir=None, revision=None, max_prefill_token_num=8192, num_tokens_per_iter=0, max_prefill_iters=1)
gen_config = GenerationConfig(n=1, max_new_tokens=512, top_p=1.0, top_k=1, temperature=0.8, repetition_penalty=1.0, ignore_eos=False, random_seed=None, stop_words=None, bad_words=None, min_new_tokens=None, skip_special_tokens=True, logprobs=None)
pipe = pipeline("/nvme/qa_test_models/internlm/internlm2-chat-7b-w8a8",  backend_config=engine_config)
res = pipe("使用python编写一个int数组的冒泡排序代码", gen_config=gen_config)

print(res)
```

error is:

```
Traceback (most recent call last):
  File "/home/zhulin1/test.py", line 9, in <module>
    pipe = pipeline("/nvme/qa_test_models/internlm/internlm2-chat-7b-w8a8",  backend_config=engine_config)
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/api.py", line 89, in pipeline
    return pipeline_class(model_path,
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/serve/async_engine.py", line 189, in __init__
    self._build_turbomind(model_path=model_path,
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/serve/async_engine.py", line 234, in _build_turbomind
    self.engine = tm.TurboMind.from_pretrained(
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/turbomind/turbomind.py", line 340, in from_pretrained
    return cls(model_path=pretrained_model_name_or_path,
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/turbomind/turbomind.py", line 144, in __init__
    self.model_comm = self._from_hf(model_source=model_source,
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/turbomind/turbomind.py", line 235, in _from_hf
    output_model = OUTPUT_MODELS.get(output_model_name)(
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/turbomind/deploy/target_model/w4.py", line 80, in __init__
    super().__init__(input_model, cfg, to_file, out_dir)
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/turbomind/deploy/target_model/base.py", line 168, in __init__
    self.cfg = self.get_config(cfg)
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/turbomind/deploy/target_model/w4.py", line 92, in get_config
    w1s, _, _ = bin.ffn_scale(i)
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/turbomind/deploy/source_model/internlm2.py", line 139, in ffn_scale
    return ensure_fp16orint32(self._ffn(i, 'scales'))
  File "/home/zhulin1/miniconda3/envs/v5/lib/python3.10/site-packages/lmdeploy/turbomind/deploy/source_model/internlm2.py", line 62, in _ffn
    tensor = self.params[
KeyError: 'model.layers.0.feed_forward.w1.scales'
```

**Comment by lvhan028** [comment]
@zhulinJulia24, turbomind engine doesn't support w8a8. pytorch engine does

File: lmdeploy/cli/cli.py [hunk 1]
```diff
@@ -31,15 +31,15 @@ def add_parser_convert():
             description=CLI.convert.__doc__,
             help=CLI.convert.__doc__)
         # define arguments
-        parser.add_argument(
```
Comment (RunningLeon) [inline lmdeploy/cli/cli.py:34]: this is a BC for `lmdeploy convert` command, should deprecate at first and then remove later?  maybe need to strongly mentioned it in release note later.

File: lmdeploy/cli/cli.py [hunk 2]
```diff
@@ -31,15 +31,15 @@ def add_parser_convert():
...
-            'model_name',
-            type=str,
-            help='The name of the to-be-deployed model, such as llama-7b, '
-            'llama-13b, vicuna-7b and etc. You can run `lmdeploy list` to '
-            'get the supported model names')
         parser.add_argument('model_path',
                             type=str,
                             help='The directory path of the model')
+        parser.add_argument(
```
Comment (RunningLeon) [inline lmdeploy/cli/cli.py:37]: could we use `ArgumentHelper.model_name` instead.

File: lmdeploy/cli/serve.py [hunk 3]
```diff
@@ -206,7 +200,6 @@ def gradio(args):
         from lmdeploy.archs import autoget_backend
         from lmdeploy.messages import (PytorchEngineConfig,
                                        TurbomindEngineConfig)
-        from lmdeploy.model import ChatTemplateConfig
```
Comment (RunningLeon) [inline lmdeploy/cli/serve.py:209]: `PytorchEngineConfig` and `TurbomindEngineConfig` do not have `model_name`, may need to update line 219 and line 228

File: lmdeploy/cli/serve.py [hunk 4]
```diff
@@ -255,7 +242,6 @@ def gradio(args):
     def api_server(args):
```
Comment (RunningLeon) [inline lmdeploy/cli/serve.py:242]: when add `--model-name` to `lmdeploy serve api_server` with workspace turbomind model path, the server model_id does not change
```
lmdeploy serve api_server internvl2-2b-tb/ --model-name bigmodel 
2024-07-26 20:28:14,455 - lmdeploy - WARNING - internvl2-2b-tb/ seems to be a turbomind workspace, which can only be ran with turbomind engine.
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
[WARNING] gemm_config.in is not found; using default GEMM algo
HINT:    Please open http://0.0.0.0:23333 in a browser for detailed api usage!!!
HINT:    Please open http://0.0.0.0:23333 in a browser for detailed api usage!!!
HINT:    Please open http://0.0.0.0:23333 in a browser for detailed api usage!!!
```

```
curl http://0.0.0.0:23333/v1/models | grep id
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   566  100   566    0     0   184k      0 --:--:-- --:--:-- --:--:--  276k
{"object":"list","data":[{"id":"/mnt/141/InternVL2-2B/snapshots/3f4a31bf77bef03ee9e783e709f678ee13b2bfc4","object":"model","created":1721998167,"owned_by":"lmdeploy","root":"/mnt/141/InternVL2-2B/snapshots/3f4a31bf77bef03ee9e783e709f678ee13b2bfc4","parent":null,"permission":[{"id":"modelperm-FDBBQkrsKamxmiqcVRJjqJ","object":"model_permission","created":1721998167,"allow_create_engine":false,"allow_sampling":true,"allow_logprobs":true,"allow_search_indices":true,"allow_view":true,"allow_fine_tuning":false,"organization":"*","group":null,"is_blocking":false}]}]}
```
``

File: lmdeploy/cli/cli.py (hunk 1 above)
Comment (lvhan028) [reply lmdeploy/cli/cli.py:34]: Not BC any more

File: lmdeploy/cli/serve.py (hunk 4 above)
Comment (lvhan028) [reply lmdeploy/cli/serve.py:239]: Fixed

File: lmdeploy/api.py [hunk 5]
```diff
@@ -8,7 +8,6 @@
 
 
 def pipeline(model_path: str,
-             model_name: Optional[str] = None,
```
Comment (RunningLeon) [inline lmdeploy/api.py:11]: pipeline is one of the most important APIs and this BC may affect some users when they use `model_name`. Shall we deprecate and then remove later?

File: lmdeploy/cli/cli.py [hunk 6]
```diff
@@ -104,10 +102,7 @@ def add_parser_chat():
             ', "baichuan-inc/baichuan2-7b-chat" and so on')
         # common args
         ArgumentHelper.backend(parser)
-        ArgumentHelper.trust_remote_code(parser)
         # # chat template args
-        ArgumentHelper.meta_instruction(parser)
-        ArgumentHelper.cap(parser)
```
Comment (RunningLeon) [inline lmdeploy/cli/cli.py:110]: should we update the doc. it is  not included in readthedocs though.
https://github.com/InternLM/lmdeploy/blob/main/docs/en/supported_models/codellama.md#completion

File: lmdeploy/turbomind/deploy/converter.py [hunk 7]
```diff
@@ -162,28 +163,60 @@ def pack_model_repository(workspace_path: str):
...
 
 
+def get_tm_model(model_path,
+                 model_name,
+                 chat_template_name,
+                 model_format,
+                 group_size,
+                 tp,
+                 out_dir: str = None):
+    # TODO: open the following condition check in another PR,
```
Comment (RunningLeon) [inline lmdeploy/turbomind/deploy/converter.py:173]: why not open in this pr?

File: lmdeploy/turbomind/turbomind.py [hunk 8]
```diff
@@ -197,17 +155,16 @@ def _from_hf(self, model_source: ModelSource, model_path: str,
         """Load model which is in hf format."""
         assert model_source == ModelSource.HF_MODEL, \
             f'{model_source} is not supported'
+        if engine_config is None:
+            logger.warning('input engine config is None, using the defult')
```
Comment (RunningLeon) [inline lmdeploy/turbomind/turbomind.py:159]: `defult` -> `default`

File: lmdeploy/api.py (hunk 5 above)
Comment (lvhan028) [reply lmdeploy/api.py:11]: No. I already labeled this PR as BC

File: lmdeploy/cli/serve.py [hunk 9]
```diff
@@ -51,16 +51,13 @@ def add_parser_gradio():
         ArgumentHelper.download_dir(parser)
 
         # chat template args
-        ArgumentHelper.meta_instruction(parser)  # TODO remove
         ArgumentHelper.chat_template(parser)
-        ArgumentHelper.cap(parser)
```
Comment (AllentDan) [inline lmdeploy/cli/serve.py:56]: may update codellama document https://github.com/InternLM/lmdeploy/blob/main/docs/zh_cn/supported_models/codellama.md

File: lmdeploy/cli/utils.py [hunk 10]
```diff
@@ -222,18 +245,6 @@ def repetition_penalty(parser):
...
-    @staticmethod
-    def cap(parser):
-        """Add argument cap to parser."""
-
-        return parser.add_argument(
-            '--cap',
-            type=str,
-            default='chat',
-            choices=['completion', 'infilling', 'chat', 'python'],
-            help='The capability of a model. '
```
Comment (AllentDan) [inline lmdeploy/cli/utils.py:234]: May update https://github.com/InternLM/lmdeploy/blob/main/autotest/utils/run_client_chat.py#L32

File: lmdeploy/cli/serve.py (hunk 9 above)
Comment (lvhan028) [reply lmdeploy/cli/serve.py:56]: Fixed

File: lmdeploy/cli/utils.py (hunk 10 above)
Comment (lvhan028) [reply lmdeploy/cli/utils.py:234]: @zhulinJulia24 I remove `--cap` from the CLI chat and use `--chat-template <a json file>` instead.
Could you help fixing the test cases?

Comment (zhulinJulia24) [reply lmdeploy/cli/utils.py:234]: > @zhulinJulia24 I remove `--cap` from the CLI chat and use `--chat-template <a json file>` instead. Could you help fix the test cases?

already fixed in pr https://github.com/InternLM/lmdeploy/pull/2183

File: lmdeploy/turbomind/turbomind.py [hunk 11]
```diff
@@ -331,18 +277,16 @@ def from_pretrained(cls,
                       and so on.
             model_format (str): model format
             group_size (int): group size
```
Comment (irexyc) [inline lmdeploy/turbomind/turbomind.py:279]: shoud be removed.

File: lmdeploy/turbomind/deploy/converter.py [hunk 12]
```diff
@@ -197,11 +230,17 @@ def main(model_name: str,
...
+        logger.warning(
+            'The argument `<model_name>` is deprecated and unused now. '
+            'It will be removed on 2024.12.31. It was originally used to '
+            'specify the name of the built-in chat template, but now it '
+            'is substituted with a clearer parameter `--chat-template`')
+    if chat_template is None:
+        chat_template = best_match_model(model_path)
+    assert chat_template in MODELS.module_dict.keys(), \
+        f"chat template '{chat_template}' is not a built-in template. " \
+        f'The built-ins are: {MODELS.module_dict.keys()}'
```
Comment (irexyc) [inline lmdeploy/turbomind/deploy/converter.py:243]: If the user doesn't provide chat_template and the best_match_model return base, should we treat this as an error?

File: benchmark/profile_generation.py [hunk 13]
```diff
@@ -78,14 +78,12 @@ def warmup(model, concurrency: int, input_ids: List[int], warmup_round: int,
         return
 
     print('start to warmup ...')
-    output_seqlen = gen_config.max_new_tokens
```
Comment (RunningLeon) [inline benchmark/profile_generation.py:81]: `python benchmark/profile_generation.py /nvme/shared_data/InternLM/internlm2_5-7b-chat --backend pytorch`
--- 
profiling ... concurrency: 1, n_prompt_token: 1, n_completion_token: 128, test_round: 3, warmup_round: 1
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████| 8/8 [00:05<00:00,  1.58it/s]
start to warmup ...
end warmup, elapsed time: 5.39s
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:07<00:00,  2.49s/it]

--------------------------------------------------
total time: 7.46s
concurrency: 1, test_round: 3
input_tokens: 1, output_tokens: 128
first_token latency(min, max, ave): 0.041s, 0.041s, 0.041s
total_token latency(min, max, ave): 2.462s, 2.523s, 2.496s
token_latency percentiles(50%,75%,95%,99%)(s): [0.019, 0.019, 0.037, 0.04]
throughput(output): 51.46 token/s
throughput(total): 51.86 token/s
--------------------------------------------------
Traceback (most recent call last):
  File "benchmark/profile_generation.py", line 487, in <module>
    main()
  File "benchmark/profile_generation.py", line 442, in main
    output = _process_map(profile_target, (args.model_path, ))
  File "benchmark/profile_generation.py", line 385, in _process_map
    raise ret
AttributeError: 'Engine' object has no attribute 'model_name'

File: lmdeploy/turbomind/deploy/converter.py (hunk 12 above)
Comment (lvhan028) [reply lmdeploy/turbomind/deploy/converter.py:243]: Let's keep it unchanged in the PR. We can discuss about it internally later on

**Review by RunningLeon — APPROVED** [review]
LGTM

**Review by AllentDan — APPROVED** [review]
LGTM

**Pull request merged**